### PR TITLE
add support for 'COMPOSER' env variable for alternate name of 'composer.json'

### LINKED ISSUE #4213: Psalm doesn't follow "COMPOSER" env variable for alternate name of "composer.json"
If the `$COMPOSER` env variable is set, then it is possible to set the filename of `composer.json` to something else. So the [composer manual says](https://getcomposer.org/doc/03-cli.md#composer).

But `psalm`(`psalm/phar`) seems to stick to the `composer.json` when initialize with`psalm --init`. And throws the `could not find composer.json during initialization.` error as below.

```shellsession
$ ./vendor/bin/psalm.phar --init --root=src
Calculating best config level based on project files
Uncaught Psalm\Exception\ConfigCreationException: Problem during config autodiscovery - could not find composer.json during initialization. in 
...
```

Is there a way to specify the alt name of composer.json?

## How to reproduce

- Before `composer install`

```shellsession
$ tree
.
├── composer.dev.json
├── composer.json
└── src
    └── main.php

1 directory, 3 files
```
```shellsession
$ cat composer.json
{
    "name": "keinos/sample",
    "desctiption": "sample"
}
```
```shellsession
$ cat composer.dev.json
{
    "name": "keinos/sample",
    "desctiption": "sample",
    "require-dev": {
        "psalm/phar": "^3.16"
    }
}
```
```php
$ cat ./src/main.php
<?php

echo 'Hello, world!', PHP_EOL;

```

- After `composer install` with `composer.dev.json`

```shellsession
$ # Set COMPOSER env var as dev mode 
$ export COMPOSER='composer.dev.json'
$ echo $COMPOSER
composer.dev.json
$
$ # Install (Same as: $ COMPOSER='composer.dev.json' composer install )
$ composer install
Loading composer repositories with package information
Updating dependencies (including require-dev)
Package operations: 1 install, 0 updates, 0 removals
  - Installing psalm/phar (3.16): Loading from cache
Writing lock file
Generating autoload files
```
```shellsession
$ tree
.
├── composer.dev.json
├── composer.dev.lock
├── composer.json
├── src
│   └── main.php
└── vendor
    ├── autoload.php
    ├── bin
    │   └── psalm.phar -> ../psalm/phar/psalm.phar
    ├── composer
    │   ├── ClassLoader.php
    │   ├── LICENSE
    │   ├── autoload_classmap.php
    │   ├── autoload_namespaces.php
    │   ├── autoload_psr4.php
    │   ├── autoload_real.php
    │   ├── autoload_static.php
    │   └── installed.json
    └── psalm
        └── phar
            ├── README.md
            ├── composer.json
            └── psalm.phar

6 directories, 17 files
```
```shellsession
$ # Initialize psalm
$ ./vendor/bin/psalm.phar --init --root=src
Calculating best config level based on project files
Uncaught Psalm\Exception\ConfigCreationException: Problem during config autodiscovery - could not find composer.json during initialization. in phar:///Volumes/php_test/psalm_autoload/vendor/psalm/phar/psalm.phar/src/Psalm/Config/Creator.php:123
Stack trace:
#0 phar:///Volumes/php_test/psalm_autoload/vendor/psalm/phar/psalm.phar/src/Psalm/Config/Creator.php(50): Psalm\Config\Creator::getPaths('/Volumes/php_te...', NULL)
#1 phar:///Volumes/php_test/psalm_autoload/vendor/psalm/phar/psalm.phar/src/Psalm/Config/Creator.php(58): Psalm\Config\Creator::getContents('/Volumes/php_te...', NULL, 1, 'vendor')
#2 phar:///Volumes/php_test/psalm_autoload/vendor/psalm/phar/psalm.phar/src/psalm.php(254): Psalm\Config\Creator::createBareConfig('/Volumes/php_te...', NULL, 'vendor')
#3 phar:///Volumes/php_test/psalm_autoload/vendor/psalm/phar/psalm.phar/psalm(5): require_once('phar:///Volumes...')
#4 /Volumes/php_test/psalm_autoload/vendor/psalm/phar/psalm.phar(14): require('phar:///Volumes...')
#5 {main}
(Psalm 3.16@d03e5ef057d6adc656c0ff7e166c50b73b4f48f3 crashed due to an uncaught Throwable)

$
$ COMPOSER='composer.dev.json' ./vendor/bin/psalm.phar --init --root=src
Calculating best config level based on project files
Uncaught Psalm\Exception\ConfigCreationException: Problem during config autodiscovery - could not find composer.json during initialization. in phar:///Volumes/php_test/psalm_autoload/vendor/psalm/phar/psalm.phar/src/Psalm/Config/Creator.php:123
Stack trace:
#0 phar:///Volumes/php_test/psalm_autoload/vendor/psalm/phar/psalm.phar/src/Psalm/Config/Creator.php(50): Psalm\Config\Creator::getPaths('/Volumes/php_te...', NULL)
#1 phar:///Volumes/php_test/psalm_autoload/vendor/psalm/phar/psalm.phar/src/Psalm/Config/Creator.php(58): Psalm\Config\Creator::getContents('/Volumes/php_te...', NULL, 1, 'vendor')
#2 phar:///Volumes/php_test/psalm_autoload/vendor/psalm/phar/psalm.phar/src/psalm.php(254): Psalm\Config\Creator::createBareConfig('/Volumes/php_te...', NULL, 'vendor')
#3 phar:///Volumes/php_test/psalm_autoload/vendor/psalm/phar/psalm.phar/psalm(5): require_once('phar:///Volumes...')
#4 /Volumes/php_test/psalm_autoload/vendor/psalm/phar/psalm.phar(14): require('phar:///Volumes...')
#5 {main}
(Psalm 3.16@d03e5ef057d6adc656c0ff7e166c50b73b4f48f3 crashed due to an uncaught Throwable)

```

## Note

1. This is a simplifyed sample. But in our project, we have two composer config files; `composer.json` for production and `composer.dev.json` for development/testing/static analyze/etc.
2. Placing `psalm.xml` manually will cause `Could not find any composer autoloaders in /Volumes/php_test/psalm_autoload/` error.
3. The below workaround will fix this issue but thats our very last choice. 😭 
    1. Include the "`require-dev`" in "`composer.json`"  and not use `composer.dev.json` 
    2. Rename the `composer.dev.lock` to `composer.lock`

## Env info

- OS: macOS Catalina (OSX 10.15.6)
- Psalm: v3.16
- PHP: v7.3.11 (CLI)

### PR DESCRIPTION
**Pull request by azjezz**
closes #4213

### CODE REVIEW

**Comment by weirdan** [comment]
According to composer docs, setting `COMPOSER` also affects the lock file name:
> For example:
> 
>     COMPOSER=composer-other.json php composer.phar install
>
> The generated lock file will use the same name: **`composer-other.lock`** in this example.

**Comment by azjezz** [comment]
`vendor/` seems to be hardcoded in multiple places as well, should we add support for that?

**Comment by azjezz** [comment]
RFR

File: composer.json [hunk 1]
```diff
@@ -76,6 +76,7 @@
         },
         "files": [
             "src/functions.php",
+            "src/command_functions.php",
```
Comment (azjezz) [inline composer.json:79]: The reason behind this is that the newly added functions are used both in binaries ( before including autoload.php ) and within internal classes.

without this, psalm would work as excepted, however, if someone is working on a project that extends psalm itself by using it as a non-dev dependency and using `Psalm\Config`, they would need the newly added functions as well.

Comment (azjezz) [reply composer.json:79]: Hm, not sure what the solution is here, currently breaking as composer requires this using `require` instead of `require_once`

Comment (azjezz) [reply composer.json:79]: @weirdan do you think redeclaring the functions twice is the way to go here? I can't come up with another solution ...

Comment (weirdan) [reply composer.json:79]: I would probably suggest to add those functions to an autoloadable class (as a static method), and `require_once` that file in binaries (or `command_functions`).

Comment (azjezz) [reply composer.json]: I have added an internal class ( `Psalm\Internal\Compposer` ), which we could use later to add more functions that can take care of tracking down `vendor` and `autoload.php`.

**Comment by azjezz** [comment]
IMHO, the architect of the commands is broken, psalm shouldn't have to deal with anything that is in `command_functions.php`

**Comment by azjezz** [comment]
appveyor failure is unrelated.

**Comment by muglug** [comment]
Thanks!

**Comment by KEINOS** [comment]
Thank you! Thank you! Thank you! @azjezz You nailed it!

**Pull request merged**